core/src/main/java/kafka/server/share/SharePartitionManager.java

```
@@ -603,22 +612,43 @@ private void maybeCompleteInitializationWithException(
+        log.error("Error initializing share partition with key {}", sharePartitionKey, throwable);
+        future.completeExceptionally(throwable);
+    }
+
+    private void handleSharePartitionException(
+        SharePartitionKey sharePartitionKey,
+        Throwable throwable
+    ) {
         if (throwable instanceof NotLeaderOrFollowerException || throwable instanceof FencedStateEpochException) {
             log.info("The share partition with key {} is fenced: {}", sharePartitionKey, throwable.getMessage());
             // The share partition is fenced hence remove the partition from map and let the client retry.
             // But surface the error to the client so client might take some action i.e. re-fetch
             // the metadata and retry the fetch on new leader.
-            partitionCacheMap.remove(sharePartitionKey);
-            future.completeExceptionally(throwable);
-            return;
+            partitionCacheMap.computeIfPresent(sharePartitionKey, (k, v) -> null);
         }
+    }
 
-        // The partition initialization failed, so complete the request with the exception.
-        // The server should not be in this state, so log the error on broker and surface the same
-        // to the client. As of now this state is in-recoverable for the broker, and we should
-        // investigate the root cause of the error.
-        log.error("Error initializing share partition with key {}", sharePartitionKey, throwable);
-        future.completeExceptionally(throwable);
+    // TODO: Should the return be -1 or throw an exception?
+    private int getLeaderEpoch(TopicPartition tp) {
+        Either<Errors, Partition> partitionOrError = replicaManager.getPartitionOrError(tp);
+        if (partitionOrError.isLeft()) {
+          log.error("Failed to get partition leader for topic partition: {}-{} due to error: {}",
+              tp.topic(), tp.partition(), partitionOrError.left().get());
+          return -1;
```

We need to return a NOT_LEADER_OR_FOLLOWER error to the client if the broker is not the leader.

---

Done.